cmd/flux-job: Output key options for job info

chu11 opened the pull request:
Per idea from @SteVwonder  in #2851, when a key is not specified for flux job info, output possible keys by going through the list of KVS keys in the KVS job path.

I thought this was going to be super simple, so I was just going to program this up while I got woken up by the baby at 3am.  But it ended up being a tad trickier than I was hoping for.

Basically, there's always the chance that when you're going through all the possible KVS keys in the job path, there's a chance that the guest namespace will disappear and be re-absorbed back into the primary namespace.  To solve this, if I get a ENOTSUP indicating that the namespace has disappeared, I just re-do the entire KVS search again.  This leads to the code being more annoying that I'd like it to be.

Maybe it's overkill.  Instead, I could have just accepted the rare error when there is a walk through a KVS directory and the guest namespace disappears on you.

chu11 commented:
re-pushed, build error in travis

SteVwonder commented on t/t2231-job-info-lookup.t:
```diff
@@ -33,6 +33,37 @@ test_expect_success 'job-info: generate jobspec for simple test job' '
+test_expect_success 'flux job info listing of keys works (job running)' '
+        jobid=`flux job submit sleeplong.json` &&
+        fj_wait_event $jobid start >/dev/null &&
+	flux job info $jobid > list_kesy2.out &&
```
Nitpick:
```suggestion
	flux job info $jobid > list_keys2.out &&
```

chu11 replied:
oops, thanks for the catch.  Re-pushed with a fix.  Since its just spelling errors, just squashed the fix.

grondo commented:
Just taking a look at this.

My first comment is that in the default output for `flux job info ID`, it might be useful to issue a message similar to "Usage", e.g. instead of just dumping the list of available keys something like this:
```
$ flux job info ƒ5yQZHdh
Usage: flux job info ID KEY...
Possible values for KEY include:
J
R
eventlog
guest.exec.eventlog
guest.input
guest.output
jobspec
```

Depending on preference, the "Usage: " string could be dropped and just the message "Possible values for KEY: "...

chu11 commented:
@grondo now that you mention it, that's something obvious I should have added :P 

Re-pushed and rebased.  I added the output "Missing lookup key(s), possible keys:" before the list of potential keys.

garlick reviewed (changes requested):
Guests don't have direct access to the primary namespace, so the KVS walk is only going to work for the instance owner.

Another drawback of dynamically obtaining the list of keys for a Usage message is that a given job may not have all possible keys yet.  For example it won't have _R_ or a guest namespace until it starts running.

It seems to me that it would make more sense here to just dump a static list of "some possible keys" that includes the ones defined in RFC 16 as well as the generally known ones from the guest namespace.  If the user tries one and gets ENOENT, they can reason "oh right, the job doesn't have resources yet so of course it doesn't have R".  But leaving well known keys off of the usage list (or worse: permission denied) kind of leaves them in the dark.

Before I remembered the above I was going to suggest also:
* usage messages should go to stderr not stdout
* is printing of an error message suppressed in the ENOTSUP case?
* should building of Usage output ever print error messages?  Seems like it should just be best effort and the `errstr` handing may be over the top.

chu11 commented:
> Guests don't have direct access to the primary namespace, so the KVS walk is only going to work for the instance owner.

Ahh, good point.  I had forgotten about that.

> Another drawback of dynamically obtaining the list of keys for a Usage message is that a given job may not have all possible keys yet. For example it won't have R or a guest namespace until it starts running.

Yeah, this was meant as a general "best case" approach.  Accepting that it may not be complete at that point in time.

> It seems to me that it would make more sense here to just dump a static list of "some possible keys" that includes the ones defined in RFC 16 as well as the generally known ones from the guest namespace. If the user tries one and gets ENOENT, they can reason "oh right, the job doesn't have resources yet so of course it doesn't have R". But leaving well known keys off of the usage list (or worse: permission denied) kind of leaves them in the dark.

Perhaps my implementation was over the top.  Lets try this approach and see how it works out.

chu11 commented:
rebased & re-pushed, this time doing a far simpler implementation, just hard coding a list of the 7 most common keys that users might what to request:

```
Missing lookup key(s), common keys:
J
R
eventlog
jobspec
guest.exec.eventlog
guest.input
guest.output

```

chu11 commented:
re-pushed to get a smidge more coverage by covering the `flux job info` without specifying a jobid case.

garlick commented on src/cmd/flux-job.c:
```diff
@@ -2713,6 +2713,19 @@ int cmd_wait_event (optparse_t *p, int argc, char **argv)
     return (0);
 }
 
+void info_list_keys (flux_t *h)
```
Perhaps drop the unused flux_t argument and rename function to info_usage()?
O/w LGTM

chu11 replied:
oh duh, yeah, i don't use that arg anymore :-)

chu11 commented:
re-pushed with @garlick comment

garlick commented:
Thanks!  I think you can mwp when ready.
